src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -703,15 +703,17 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
                     // parsed documents over to the new project state so we can preserve as much information as
                     // possible.
 
-                    using var _1 = ArrayBuilder<DocumentState>.GetInstance(out var documentsWithTrees);
-                    using var _2 = ArrayBuilder<SyntaxTree>.GetInstance(out var alreadyParsedTrees);
+                    var alreadyParsedCount = this.ProjectState.DocumentStates.States.Count(s => s.Value.TryGetSyntaxTree(out _));
+
+                    using var _1 = ArrayBuilder<DocumentState>.GetInstance(alreadyParsedCount, out var documentsWithTrees);
```

Comment:
>using var _1 = ArrayBuilder<DocumentState>.GetInstance(alreadyParsedCount, out var documentsWithTrees); [](http://example.com/codeflow?start=20&length=103)

I know you like to keep the concept count simple, but it seems like since the size is known, you might as well use ImmutableArray.Builder directly as there's no advantage to churning the arraypool here.